ibtk/src/utilities/MarkerPatchHierarchy.cpp

```diff
@@ -343,6 +428,242 @@ MarkerPatchHierarchy::reinit(const EigenAlignedVector<IBTK::Point>& positions,
+    const int ierr = MPI_Allreduce(
+        MPI_IN_PLACE, marker_check.data(), marker_check.size(), MPI_UNSIGNED, MPI_SUM, IBTK_MPI::getCommunicator());
+    TBOX_ASSERT(ierr == 0);
+    for (unsigned int k = 0; k < getNumberOfMarkers(); ++k)
+    {
+        if (marker_check[k] != 1)
+        {
+            TBOX_ERROR(d_object_name
+                       << ": Marker point " << k << " is presently owned by " << marker_check[k]
+                       << " patches. The most likely cause of this error is that the CFL number is greater than 1.");
+        }
+    }
+#endif
+}
+
+void
+MarkerPatchHierarchy::putToDatabase(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db)
+{
+    TBOX_ASSERT(d_num_markers <= std::numeric_limits<int>::max());
+    db->putInteger("num_markers", int(d_num_markers));
+
+    auto put_marker_patch = [&](const MarkerPatch& marker_patch, const std::string& prefix)
+    {
+        db->putInteger(prefix + "_num_markers", static_cast<int>(marker_patch.d_indices.size()));
+        // Yet another SAMRAI bug: we are not allowed to store zero-length
+        // arrays in the database so we have to special-case that here ourselves
+        if (marker_patch.d_indices.size() > 0)
+        {
+            db->putIntegerArray(
+                prefix + "_indices", marker_patch.d_indices.data(), static_cast<int>(marker_patch.d_indices.size()));
+            db->putDoubleArray(prefix + "_positions",
+                               marker_patch.d_positions.data(),
+                               static_cast<int>(marker_patch.d_positions.size()));
+            db->putDoubleArray(prefix + "_velocities",
+                               marker_patch.d_velocities.data(),
+                               static_cast<int>(marker_patch.d_velocities.size()));
+        }
+    };
+
+    int marker_patch_num = 0;
+    for (const auto& level_marker_patches : d_marker_patches)
+    {
+        for (const auto& marker_patch : level_marker_patches)
+        {
+            const std::string key_prefix = "patch_" + std::to_string(marker_patch_num);
+            put_marker_patch(marker_patch, key_prefix);
+            ++marker_patch_num;
+        }
+    }
+
+    put_marker_patch(d_markers_outside_domain, "outside_domain_");
+}
+
+void
+MarkerPatchHierarchy::writeH5Part(const std::string& filename,
+                                  const int /*time_step*/,
+                                  const double simulation_time,
+                                  const bool write_velocities) const
+{
+    const auto pair = collectAllMarkers();
+    const auto& positions = pair.first;
+    const auto& velocities = pair.second;
+
+    const std::array<std::string, NDIM> position_datasets{ { "x",
+                                                             "y",
+#if NDIM == 3
+                                                             "z"
+#endif
+    } };
+
+    // technically momentum, but marker points don't have mass so this is the
```

Comment:
Why are you using the momentum nomenclature --- is it required by VisIt?

Reply:
Yes - H5Part stores momenta, not velocity. I'll add a comment.